sdk/python/offlinehub.py

```diff
@@ -0,0 +1,163 @@
+
+        Args:
+            embedding_tuples: A dictionary from key to embedding or an iterator
+            of key, embedding pairs where key is a string and embedding is a
+            python list.
+        """
+        embeddings = []
+        idxs = []
+        if isinstance(embedding_tuples, Mapping):
+            embedding_tuples = embedding_tuples.items()
+
+        for key, embedding in embedding_tuples:
+            embeddings.append(embedding)
+            idxs.append(self._mapper.to_idx(key))
+            self._data[key] = embedding
+        # add_items fails if you give it two empty arrays.
+        if len(idxs) == 0:
+            return
+        self._add_capacity_to_fit(len(idxs))
+        self._idx.add_items(embeddings, idxs)
+
+    def multiget(self, keys):
+        """Get multiple embeddings at once.
+
+        Args:
+            keys: An iterator of embedding indices for retrieval.
+
+        Returns:
+            A list of embeddings.
+        """
+        return [self._data[key] for key in keys]
+
+    def nearest_neighbor(self, key, num):
+        """Finds N nearest neighbors for a given embedding record.
+
+        Args:
+            key: The embedding index to get the nearest neighbors of.
+            num: The number of nearest neighbors.
+
+        Returns:
+            A num size list of embedding vectors that are closest to the
+            provided embedding.
+        """
+        idx = self._mapper.to_idx(key)
+        # knn_query expects a list of indices to work in batch.
+        idxs = [idx]
+        # knn_query returns itself, so we get one extra nearest neighbor and drop
+        # idx when mapping the indices.
+        list_results, list_distances = self._idx.knn_query(idxs, num + 1)
+        results = list_results[0]
+        return [
+            self._mapper.to_key(result) for result in results if result != idx
+        ]
+
+    def _add_capacity_to_fit(self, size):
+        min_cap = self._size + size
+        if min_cap > self._cap:
+            self._cap *= 2
+            self._idx.resize_index(self._cap)
+
+
+class HnswlibIndexMapper:
+    """ Maps keys to atomically increasing ints
+
+    hnswlib.Index expects labels to be atomically increasing ints, but our API
+    handles any type of key. This class is used for mapping back and forth
+    """
+
+    def __init__(self):
+        self._key_to_idx = {}
+        self._idx_to_key = {}
+        self._next_idx = 0
+
+    def to_idx(self, key):
+        """ Turn a key into an hnswlib index
+
+        Args:
+            key: The key to transform
+
+        Returns:
+            An index to be used in hnswlib
+        """
+        if key in self._key_to_idx:
+            return self._key_to_idx[key]
+        else:
+            idx = self._next_idx
+            self._next_idx += 1
+            self._key_to_idx[key] = idx
+            self._idx_to_key[idx] = key
+            return idx
+
+    def to_key(self, idx):
```

Comment:
perhaps include error handling if index doesn't exist

Reply:
What should it do you think? Currently it would just throw an error due to the map

Reply:
Would depend on the use case but you're right, Python will just raise a key error and reference the line. Just was unsure about needed robustness of error handling conventions.

Reply:
I think for Python this is fine for now, we could catch and wrap it in our own error or provide another API for a safer to_key that won't throw in the future